modules/sor/lib/poolsV3/liquidityBootstrapping/liquidityBootstrapping.ts

```
@@ -0,0 +1,251 @@
+        this.endTime = endTime;
+        this.startWeights = startWeights;
+        this.endWeights = endWeights;
+        this.poolState = this.getPoolState(hookState?.hookType);
+    }
+
+    public getPoolState(hookName?: string): LiquidityBootstrappingState {
+        const poolState: LiquidityBootstrappingState = {
+            poolType: 'LIQUIDITY_BOOTSTRAPPING',
+            poolAddress: this.address,
+            tokens: this.tokens.map((t) => t.token.address),
+            scalingFactors: this.tokens.map((t) => t.scalar),
+            tokenRates: this.tokens.map((t) => ('rate' in t ? t.rate : WAD)),
+            balancesLiveScaled18: this.tokens.map((t) => t.scale18),
+            swapFee: this.swapFee,
+            aggregateSwapFee: this.aggregateSwapFee,
+            totalSupply: this.totalShares,
+            weights: this.tokens.map((t) => t.weight),
+            supportsUnbalancedLiquidity: !this.liquidityManagement.disableUnbalancedLiquidity,
+            projectTokenIndex: this.projectTokenIndex,
+            isProjectTokenSwapInBlocked: this.isProjectTokenSwapInBlocked,
+            isSwapEnabled: this.isSwapEnabled,
+            currentTimestamp: this.currentTimestamp,
+            startTime: this.startTime,
+            endTime: this.endTime,
+            startWeights: this.startWeights,
+            endWeights: this.endWeights,
+        };
+
+        poolState.hookType = hookName;
+
+        return poolState;
+    }
+
+    public swapGivenIn(tokenIn: Token, tokenOut: Token, swapAmount: TokenAmount): TokenAmount {
+        // TODO: Instead of throwing an error, we could return a TokenAmount with 0 amount?
+        // has the LBP started already? Indicated by swapEnabled
+        if (!this.isSwapEnabled) {
+            throw new Error('LBP has not started yet');
+        }
+
+        // project token swap in can be blocked so check for it here
+        if (this.isProjectTokenSwapInBlocked && tokenIn.isSameAddress(this.projectToken)) {
```

Isn't these checks already being handled within balancer-maths? Any benefit on splitting the logic between here and there?

---

You are correct, that both of these checks are done in balancer maths. I can remove them but wanted to make sure, I understand the SOR correctly. 
- If I throw an error, the whole "SOR route fetching process" errors out and should propagate the error up.

I could however - in difference to math - return 0 amounts out here to make the getPathsWithPools function continue executing. Therefor I think it could be better to leave these snippets in and instead return TokenAmount.fromRawAmount(tokenIn/tokenOut , 0) basically. What do you think?

---

Resolved this topic in our daily standup. I wasn't aware of the relevant try catch statement in the code later. As part of 9629dbfe203f002cd90d900604ed21570ba7da25

---

Addressed. With add liquidity/remove liquidity checks being done in the basePool and other verifications being done in balancer maths, I could actually remove the implementation of `swapGivenIn` & `swapGivenOut` to be handled by the Basepool. Thus refactored the pool itself quite a bit.